Added API request functions

ggobugi27 opened the pull request:
Added following functions: 
 - simpleLocalCommand: 
       returns Pact command in JSON format of local API request. 
 - sendCommand: 
       makes request to /send, then requests to /listen, then returns responseJSON.result of listen. 
 - sendLocalCommand: 
       makes request to /local and returns responseJSON.result

sirlensalot commented on pact-lang-api.js:
```diff
@@ -212,14 +212,20 @@ var enforceArray = function(val,msg) {
+var simpleLocalCommand = function(keyPairs, nonce, pactCode, envData, meta) {
+  return prepareExecCmd(keyPairs, nonce, pactCode, envData, meta);
+};
+
```
This is a duplicate of `prepareExecCmd`. We can look at a renaming of that function if you want but we don't need a dupe.

ggobugi27 replied:
Removed the dupe, but exported it as Pact.simple.exec.createLocalCommand to keep consistent with create...Command functions.

sirlensalot commented on pact-lang-api.js:
```diff
@@ -292,20 +296,70 @@ var mkExp = function(pgmName) {
+ * @property apiHost {string} - host running Pact server
+ * @return {object} - tx result received from pact server.
+ */
+const sendCommand = async function ({pactCode, keyPairs, nonce=new Date().toISOString(), envData={}, meta=mkMeta("","",0,0), apiHost}) {
```
Let's avoid repeating all the arguments of `simpleExecCommand` and instead just take the result from that as an `object` argument.

sirlensalot replied:
Also let's break this into two functions, one for just the send `await fetch`, then one for listen, and let's add one for poll.

ggobugi27 replied:
- Separated send/listen, added poll. -  fetchSend, fetchLoccal, fetchListen, fetchPoll. 
- About arguments, this will be the first function that users call, which then calls simpleExecCommand with its arguments.

sirlensalot commented on pact-lang-api.js:
```diff
@@ -329,8 +383,11 @@ module.exports = {
     }
-  }
+  },
+  sendCommand: sendCommand,
```
let's follow the pattern above and put these in some group, like `fetch` or `api` or something.

ggobugi27 replied:
Grouped the functions into Pact.fetch.send, Pact.fetch.local, Pact.fetch.poll, Pact.fetch.listen

sirlensalot commented on pact-lang-api.js:
```diff
@@ -52,7 +52,7 @@ var hash = function(s) {
 var base64UrlEncode = function(s) {
   return base64url.escape(base64url.encode(s));
-}
+};
```
do we need these or not? asking b/c on line 213 below it looks like there isn't a semicolon. We should standardize and do all or none, your call.

ggobugi27 replied:
Removed all semicolons from the function closure brackets.

slpopejoy commented on pact-lang-api.js:
```diff
@@ -292,19 +289,98 @@ var mkExp = function(pgmName) {
+
+/**
+ * Sends Pact command to a running Pact server and retrieves tx result.
+ * @param {Object}
```
We need a `@typedef` for the command object documented here and `fetchLocal`, so we can simply say `@param [ref-to-command-typedef]`. There's some way to introduce a typedef that can be referenced elsewhere, I can't find it right now.

sirlensalot commented on pact-lang-api.js:
```diff
@@ -292,19 +289,98 @@ var mkExp = function(pgmName) {
+
+/**
+ * Sends Pact command to a running Pact server and retrieves tx result.
+ * @param {Object}
```
We need a `@typedef` for the command object documented here and `fetchLocal`, so we can simply say `@param [ref-to-command-typedef]`. There's some way to introduce a typedef that can be referenced elsewhere, I can't find it right now.

ggobugi27 replied:
Added @typedef {Object} execCmd, and added @param{execCmd} on fetchSend, fetchLocal. For todomvc, I've checked the 4(send, local, poll, listen) API requests working, but haven't updated the whole app - will work on it.

ggobugi27 commented:
* Changed fetchSend, fetchLocal, fetchPoll, fetchListen to accept (cmd, apiHost)
   - fetchSend can receive a command object or a list of multiple command objects. 
* Added default nonce with current date and default empty meta to prepareExecCmd
* Edited docstrings and updated readme. 
@slpopejoy Please review the changes. These work with todomvc's create-react-app branch.

sirlensalot commented on readme.md:
```diff
@@ -44,25 +44,120 @@ A helper function for constructing native Pact commands.
+  ex:
+    const cmds = [{
+                    keyPairs: KEY_PAIR,
+                    pactCode: "todos.delete-todos "id-1""
```
these need escaped quotes or single quotes around the whole thing

ggobugi27 replied:
Changed to Single quotes

sirlensalot commented on readme.md:
```diff
@@ -44,25 +44,120 @@ A helper function for constructing native Pact commands.
+Pact.fetch.poll({requestKeys: ["..."]}, <apiHost:string>) -> [{requestKey: "...", result: {...}}, ...]
+
+  ex:
+    const cmd = { requestKeys: [ "6ue-lrwXaLcDyxDwJ1nuLzOfFtnQ2TaF0_Or_X0KnbE",
```
It might be better to show an example of using request keys from a previously sent command(s)

ggobugi27 replied:
Used request keys from "Pact.fetch.send" example.

sirlensalot commented on pact-lang-api.js:
```diff
@@ -291,21 +348,151 @@ var mkExp = function(pgmName) {
+};
+
+/**
+<<<<<<< HEAD
```
You haven't finished merging, fix.

ggobugi27 replied:
Fixed Merging
